### ISSUE TITLE
Create or document way to clear all global memory

### ISSUE THREAD

**Issue by PMunch** [nim-lang/Nim#21403]
### Summary

This came about when working with dynamic libraries. If you write a dynamic library in Nim and then load it into a separate program there is no (documented) way to free all the memory if the loader program unloads the dynamic library. This means that if a system loads and unloads a dynamic library it will keep increasing in memory.

### Description

What I propose is to create a procedure `NimQuit` opposite of `NimMain` which will destroy all memory. Calling any other code using GC after that will be undefined behaviour. This would then be called automatically in `.fini` by attaching `__attribute((destructor))__` or hooking in `DllMain`, similar to how `NimMain` is called today. And analogous to how `--noMain` disables this behaviour `--noQuit` should prevent it from being called and `NimQuit` must be called exlicitly.

### Examples

As a simple example/testing ground I wrote this:

`loader.nim`:
```nim
import std/dynlib
type
  InitFunction = proc () {.cdecl.}
  DeinitFunction = proc () {.cdecl.}

proc loadDynamic() =
  let lib = loadLib("./liblibrary.so")
  assert lib != nil, "Error loading library"

  let init = cast[InitFunction](lib.symAddr("dynlib_init"))
  assert init != nil, "Error loading 'init' function from library"
  let deinit = cast[DeinitFunction](lib.symAddr("dynlib_deinit"))
  assert deinit != nil, "Error loading 'deinit' function from library"

  init()
  deinit()

  unloadLib(lib)

while true:
  loadDynamic()
```

`library.nim`:
```nim
var x = newSeq[byte](uint16.high)

proc NimMain() {.cdecl, importc.}

proc dynlib_init() {.cdecl, exportc, dynlib.} =
  NimMain()
  echo "Loaded dynamic library"

proc dynlib_deinit() {.cdecl, exportc, dynlib.} =
  GC_FullCollect()
```

Compiled with:
```
nim c --gc:orc loader.nim
```
and
```
nim c --app:lib --noMain --gc:orc library.nim
```

**Comment by PMunch** [comment]
After writing this I discovered `deallocHeap` in the module `gc_common`, unfortunately this isn't available for ARC/ORC it seems.

**Comment by PMunch** [comment]
Did some testing and exposing `deallocOsPages` and calling that in the `dynlib_deinit` indeed made the memory consumption stop increasing. This won't call finalizers/destructors though, which isn't great, but at least it points to `deallocHeap` as the thing that has to be implemented for ARC/ORC in order to make this possible.

**Comment by ringabout** [comment]
related https://github.com/nim-lang/Nim/issues/18262

**Comment by PMunch** [comment]
Just for reference if someone wants to give a go at implementing this. I believe that this procedure is what needs to be ported to ARC/ORC and then create a `NimExit` which basically just calls it:

https://github.com/nim-lang/Nim/blob/39d0a93d0e0bc2ebf24a9801fe49cdb71c085527/lib/system/gc_common.nim#L434-L458

**Comment by Araq** [comment]
Makes sense but this new `NimExit` won't run destructors, it simply cannot as the RTTI is not always available.

**Comment by PMunch** [comment]
Hmm, that's a bit unfortunate. But still better than just leaving all the memory dangling. Does `deallocHeap` imply `GC_FullCollect`? Let's say I have an object which wraps a file and the destructor flushes my changes. If my program ends while that object still exists the changes won't be flushed (poor design, but it's just an example). Since `deallocHeap` can't run this destructor (because it can't see which type it is) I would then have to ensure that it is freed before I exit. If I then set my reference to this object to nil and call `GC_FullCollect` this destructor should be run, but if I omit the `GC_FullCollect` call does `deallocHeap` then call this destructor? My guess is no.

**Comment by Araq** [comment]
`deallocHeap` should not call `GC_FullCollect` but `NimExit` should call it.

**Comment by JLT032** [comment]
> `deallocHeap` should not call `GC_FullCollect` but `NimExit` should call it.

is this also addressed by https://github.com/nim-lang/Nim/commit/e217bb24a1e5b92d448bf7afa2488bec89b94cfb ?

**Comment by PMunch** [comment]
Added a call to `NimDestroyGlobals` after the `GC_FullCollect` call and there was no change in behaviour.

**Comment by SpotlightKid** [comment]
@PMunch: how are you measuring the growing memory consumption?

I tried a similar test but wrote the loading program in C and I can't see a difference in the peak  memory consumption regardless of whether I load the shared lib 100 or 100_000 times when I include the `NimDestroyGlobals` call before `GC_FullCollect`, but I see an increase in peak memory consumption that directly correlates to the number of iterations of the loop when I leave out the `NimDestroyGlobals` call. I used `heaptrack` to measure.

**Comment by PMunch** [comment]
I'm just reading it out from htop, it's growing so rapidly with the above program that it is very obvious. But I'll try to run it before `GC_FullCollect` and see if it makes a difference.

**Comment by SpotlightKid** [comment]
For reference, here's my C test program: https://gist.github.com/SpotlightKid/dbe9ea0857b6a825b7b8777bea84bdd2

**Comment by PMunch** [comment]
Neither switching the order, nor running your C-based loader made any difference. I ran your loader with `./spotlight ./liblibrary.so 1000000000` and it just kept going up and up and up.

Also tried heaptrack and it says "peak RSS" was ~12Mb for 100 iterations and 7.8Gb for 100_000 iterations. This is with a liblibrary.so built from devel with only `--app:lib --noMain` as `--gc:orc` is now the default.

**Comment by PMunch** [comment]
Ran a new test with `-d:useMalloc` and now it gives 4.6Mb for both 100 and 100_000 iterations, so it seems like maybe this is a workaround. In this case it doesn't matter which order `GC_FullCollect()` and `NimDestroyGlobals` comes in.

**Comment by SpotlightKid** [comment]
Ah, yes, I do have `-d:useMalloc` in my `nim.cfg`.

**Comment by Araq** [comment]
Ah interesting, but then calling `deallocOsPages` might help with that.

**Comment by PMunch** [comment]
Yeah as I mentioned earlier it seems like `deallocOsPages`/`deallocHeap` is what needs to be exposed for this to be fixed. Then the proposed `NimQuit` can call `NimDestroyGlobals(); GC_FullCollect(); deallocOsPages();` or whatever incantation is required.

**Comment by Araq** [comment]
Ok, well, please create a PR then.
